### PR TITLE
CA-264333: use local Unix domain socket and remove username & password authentication

### PR DESCRIPTION
**Pull request by gaborigloi**

### CODE REVIEW

File: src/main.ml [hunk 1]
```diff
@@ -45,7 +45,8 @@ module SM = Storage_interface.ClientM(struct
       return (Jsonrpc.response_of_string result)
   end)
 
-let uri = ref "http://127.0.0.1/"
+(** Xapi's local Unix domain socket *)
+let uri = "file:///var/xapi/xapi"
```
Comment (lindig) [inline src/main.ml]: There is an entry `unix_domain_socket` in `xapi_globs.ml` - how are these related? If we need a new path, should this not be defined in `xapi_globs.ml`, too, rather than here?

**Review by lindig — COMMENTED** [review]
Question about where the domain socket is defined.

File: src/main.ml (hunk 1 above)
Comment (gaborigloi) [reply src/main.ml]: Good point. I've looked at the `_oasis` file of xapi, and it seems that `Xapi_globs` isn't exposed as part of a library, it's only in the `Executable xapi`. But I agree that we should have this constant in a publicly available xapi library and in `xen-api-client`.

Comment (lindig) [reply src/main.ml]: I forgot that this is outside of xapi. So maybe it is fine not to wait for xapi to expose its constants to libraries. This project, however, could have a dedicated module that contains such constants rather than spread them across the code. This will also make it easier to import them for a future library or read them from a file.

Comment (mseri) [reply src/main.ml]: We could pass it as a CLI parameter and use the value above as a default. This would give us some time to update xapi as well

**Review by mseri — COMMENTED** [review]
I think we should allow to pass the `uri` as a CLI flag parameter (and if we detect http or https keep being able to use them as well), keeping the current `Consts.xapi_unix_domain_socket_uri` as a default value

**Comment by gaborigloi** [comment]
@mseri  Done: now the URI can be configured using the CLI.

**Comment by mseri** [comment]
Does it work fine if use `--xen_api_uri http://127.0.0,1` or `--xen_api_uri http://localhost`?

File: src/main.ml [hunk 2]
```diff
@@ -198,8 +185,11 @@ let cmd =
   ] @ help in
   let port =
     let doc = "Local port to listen for connections on" in
-    Arg.(value & opt int 10809 & info [ "port" ] ~doc) in
-  Term.(ret (pure main $ port)),
+    Arg.(value & opt int Consts.standard_nbd_port & info [ "port" ] ~doc) in
+  let xen_api_uri =
+    let doc = "The URI to use when making XenAPI calls. It must point to the pool master, or to xapi's local Unix domain socket, which is the default." in
+    Arg.(value & opt string Consts.xapi_unix_domain_socket_uri & info [ "xen_api_uri" ] ~doc) in
```
Comment (lindig) [inline src/main.ml]: I think this should be "xen-api-uri" to be more consistent with how xapi and many CLI tools name flags. I would even consider a short name like "xapi" or "xapi-socket".

Comment (gaborigloi) [reply src/main.ml]: Thanks, that sounds good, I've changed the param name.

**Comment by gaborigloi** [comment]
@mseri I tested this CLI parameter with a pool of two hosts, a master and a slave, and everything worked as expected :smile: 

Worked on the master:
* `./xapi-nbd`
* `./xapi-nbd --xen-api-uri http://127.0.0.1`
* `./xapi-nbd --xen-api-uri http://localhost`
* `./xapi-nbd --xen-api-uri file:///var/xapi/xapi`

Did not work on the master:
* `./xapi-nbd --xen-api-uri file:///var/xapi/xai` -> `Caught Unix.Unix_error(Unix.ENOENT, "connect", "")`
* `./xapi-nbd --xen-api-uri http://localhostf` -> `Caught Xen_api.Failed_to_resolve_hostname("localhostf")`

Worked on the slave:
* `./xapi-nbd`
* `./xapi-nbd --xen-api-uri file:///var/xapi/xapi`
* `./xapi-nbd --xen-api-uri http://<host name of master>`
* `./xapi-nbd --xen-api-uri http://<full domain name of master>`

Did not work on the slave:
* `./xapi-nbd --xen-api-uri http://127.0.0.1`
* `./xapi-nbd --xen-api-uri http://localhost`
These failed with `Caught Server_error(HOST_IS_SLAVE, [ <IP address> ])`
So the user specifying `--xen-api-uri` has to be careful.

**Review by mseri — APPROVED** [review]

**Comment by thomassa** [comment]
All looks good to me except for the typo in the description of the overall pull-request.
